cloudinit/config/cc_grub_dpkg.py

```diff
@@ -62,22 +67,35 @@ def handle(name, cfg, _cloud, log, _args):
     idevs_empty = util.get_cfg_option_str(
         mycfg, "grub-pc/install_devices_empty", None)
 
-    if ((os.path.exists("/dev/sda1") and not os.path.exists("/dev/sda")) or
-       (os.path.exists("/dev/xvda1") and not os.path.exists("/dev/xvda"))):
+    # get the root disk where the /boot directory resides.
+    # the ESP will reside on the same root disk as /boot, but not device.
+    try:
```

Comment:
This `try` block is too large; we should have targeted exception handlers for where we specifically expect to see failures.

The contents of this try block are also a good candidate for splitting out to a separate function, I think.

Reply:
I split the contents of the try block into a separate function, `fetch_idevs()`. I also managed to remove some duplicated code when setting `idevs_empty`.
The try block is needed as calls to `util.subp()` can throw various exceptions, namely `ProcessExecutionError`.